### DISCUSSION TITLE
LittleFS FSUpload example - I'm not quite sure this is really working.

### DISCUSSION THREAD

**Discussion by fapplin** [earlephilhower/arduino-pico#746] [Q&A]
So, I tried to run the LittleFS/FSUpload example.

It would give me an output, like:

```
I have been run 12 times
---------------
```

but, I never got anything else. So, I hunted for the file "counts.txt" and it wasn't anywhere on my system. I found the "file1.txt" in my 

```
/home/frank/.arduino15/packages/rp2040/hardware/rp2040/2.3.1/libraries/LittleFS/examples/FSUpload/data/
```

but, there was no counts.txt. So, I put one in there with a value of 100. The program just increased the count to 13. So, I modified the code so that it printed a message if the was no file found (I did the same for the "file1. txt").

```
void setup() {
  Serial.begin(115200);
  delay(5000);
  LittleFS.begin();
  char buff[32];
  int cnt = 1;
  File f = LittleFS.open("counts.txt", "r");
  if (f) {
    bzero(buff, 32);
    if (f.read((uint8_t *)buff, 31)) {
      sscanf(buff, "%d", &cnt);
      Serial.printf("I have been run %d times\n", cnt);
    }
    {
      Serial.println("File not found.");
    }
    f.close();
  }

  cnt++;
  sprintf(buff, "%d\n", cnt);
  f = LittleFS.open("counts.txt", "w");
  if (f) {
    f.write(buff, strlen(buff));
    f.close();
  }

  Serial.println("---------------");
  File i = LittleFS.open("file1.txt", "r");
  if (i) {
    while (i.available()) {
      Serial.write(i.read());
    }
    Serial.println("---------------");
    i.close();
  } else {
    Serial.println("No file found.");
  }
}
```
This was the output.

```
I have been run 14 times
File not found.
---------------
No file found.
```

So, I don't think this is working the way it seems. So, the Arduino IDE creates the following files in the system /tmp directory.

![0809221113a](https://user-images.githubusercontent.com/5192768/183688641-d4b4c6f6-86ff-4685-9943-1015f893e89d.jpg)

which files are actually being installed to the Pico (I'm assuming FSUpload.mklittlefs.bin, FSUpload.ino.uf2, and FSUpload.ino.elf created in the /tmp/arduino_build_xxxxx directory)? Is this assumption correct?

Frank

**Comment by earlephilhower** [comment]
You need to upload the filesystem separately using the tool and method described here: https://arduino-pico.readthedocs.io/en/latest/fs.html#uploading-files-to-the-littlefs-file-system  .  The files in the `/tmp` directory are just app build files.

**Reply by fapplin** [reply]
I get this error every time I try.

```
Resetting /dev/ttyACM0
Converting to uf2, output size: 131072, start address: 0x101ef000
No drive to deploy.
LittleFS Upload failed! Did you close the Serial Monitor?
```

No Serial Monitor open.


I get this error whether I hold the button on the Pico and plug in the USB or just plug in the USB without holding the button.

I'm on Endeavor OS (arch based). Is there a udev that has to be defined?

**Reply by earlephilhower** [reply]
To upload the FS it uses the same 1200bps doorbell and UF2 filesystem uploads as the app upload, so if that's working then you theoretically *should* be good.

Make sure the ttyACMx is the right one, of course, and no other apps are using it (i.e. the monitor or minicom).  No need to enter bootloader mode.  I've used the FS uploader recently for the BearSSL CertStore example, but will give this example a try later today just to check.

**Reply by fapplin** [reply]
So, I get this all of the time when just trying to Upload a sketch:

```
Resetting /dev/ttyACM0
Converting to uf2, output size: 761856, start address: 0x2000
No drive to deploy.
An error occurred while uploading the sketch
```

That's why I have just been using the Boot Sel button, Export Compiled Binary, and copying the .uf2 to RPI-RP2 drive via my file manager. That's been working. Now, I need to have a small file in memory, so I thought I'd try the LittleFS library.

**Reply by earlephilhower** [reply]
If you can't upload thru the IDE, you're not going to be able to upload through the FS uploader.  Arch Linux makes it easy to configure things how you like it, but sometimes that makes it incompatible with the way other distros do it. :(

First of all, not sure if Arch supports it, but if you installed a Flatpack/Snap/etc. prepackaged version, the permissions of the app often disallow it from accessing the filesystems needed (and under Winodws it even disallows serial port access!) rendering it pretty useless.  Use the tarballs from arduino.cc if that is the case and you have a good chance of things working properly.  (There's also probably a way of adding permissions to the flatpack/snap/etc. but I've never tried since those kinds of "full requirements" packages seem to me like a recipe for DLL hell under Linux.)

The upload script is here:
https://github.com/earlephilhower/arduino-pico/blob/master/tools/uf2conv.py

and the (heroic) autodetect of path is here:
https://github.com/earlephilhower/arduino-pico/blob/11cb82b058aae7e58b43e2ac8da21fdb347f25b0/tools/uf2conv.py#L228-L279

When you hold down the BOOTSEL and plug in, is the drive auto-mounted?  What path?  Is that path in the above script, or elsewhere (so we could add it)?

**Reply by fapplin** [reply]
using  df -h, I get:

```
/dev/sdd1       128M  8.0K  128M   1% /run/media/frank/RPI-RP2
```

I'm not sure what other info. you need.

**Reply by earlephilhower** [reply]
Reading thru the code I see:
1. `/run/media` exists so `rootpath = /run/media`
2. sys/platform==linux, and `/run/media/frank` exists so `rootpath = /run/media/frank`
3. get_drives should then return a list of all dirs in `/run/media/frank` which will be `[/run/media/frank/RPI-RP2]`

My guess is that SELinux or equivalent is blocking access by the IDE to the `/run/media`.  That's the only way I can see it failing.  You'll need to look at your selinux logs to determine this, and if so add `/run/media` as part of the allow for the IDE and sub-processes.

**Reply by fapplin** [reply]
OK. I'll look into it. Thanks for all your help.

**Reply by fapplin** [reply]
Just wanted you to know that I uninstalled the Arduino IDE (which was from the repo) and re-installed it from the Arduino download site - and all is well! I can do the LittleFS Data Upload and Upload right from the Arduino IDE. Lesson learned.
